---
title: "Class19_Mini Project_Investigating Pertussis Resurgence"
author: "Changcheng (PID: A69027828)"
format: pdf
---

Pertussis (more commonly known as whooping cough) is a highly contagious respiratory disease caused by the bacterium Bordetella pertussis. People of all ages can be infected leading to violent coughing fits followed by a high-pitched intake of breath that sounds like “whoop”. Infants and toddlers have the highest risk for severe complications and death. Recent estimates from the WHO suggest that ~16 million cases and 200,000 infant deaths are due to pertussis annually 1.

1. Investigating pertussis cases by year

```{r}
#install.packages("datapasta")
library(datapasta)
```

>Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
#| echo = FALSE
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L,2020L,2021L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617,
                                          6124,2116)
       )
```

```{r}
library(ggplot2)

pertucases <- ggplot(cdc) +
  aes(x = Year, y = No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs(title = "Pertussis Cases by Year (1922-2021)", 
       y = "Number of cases")

pertucases
```

2. A tale of two vaccines (wP & aP)

>Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
pertucases + 
  geom_vline(xintercept=1946, linetype = 2, color = "blue") +
  geom_vline(xintercept=1996, linetype = 2, color = "red") +
  geom_vline(xintercept=2019, linetype = 2, color = "darkgreen") +
  geom_text(label="wP", x = 1946, y = 270000, color = "blue") +
  geom_text(label="aP", x = 1996, y = 270000, color = "red") +
  geom_text(label="Covid19", x = 2019, y = 270000, color = "darkgreen") +
  scale_y_continuous(labels = scales::number)
```

>Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

Pertussis cases are increasing again after the introduction of aP vaccine.
A possible explanation is that Bordetella pertussis has evolved to escape from vaccine immunity and we need novel types of vaccines.

3. Exploring CMI-PB data

```{r}
# Allows us to read, write and process JSON data
#install.packages("jsonlite")
library(jsonlite)
```

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

```{r}
head(subject, 3)
```

>Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

There are 60 aP and 58 wP infancy vaccinated subjects

>Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex) 
```

There are 39 Males and 79 Females

>Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$biological_sex, subject$race)
```

> Q, Make a histogram of the subject age distribution and facet by infancy_vac

Side-Note: Working with dates

```{r}
#| echo = FALSE
#install.packages("lubridate")
library(lubridate)
```

```{r}
today()
today() - ymd("2000-01-01")
time_length(today() - ymd("2000-01-01"), "years")
```
>Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

(i)
```{r}
subwP <- subject$infancy_vac == "wP"
subAge <- time_length(today() - ymd(subject$year_of_birth), "years")
AvgwPAge <- sum(subwP * subAge) / sum(subwP)
AvgwPAge
```
The average age of wP individuals is 36.3 years.

(ii)
```{r}
subaP <- subject$infancy_vac == "aP"
subAge <- time_length(today() - ymd(subject$year_of_birth), "years")
AvgaPAge <- sum(subaP * subAge) / sum(subaP)
AvgaPAge
```
The average age of wP individuals is 26.0 years.

```{r}
wP1 <- subwP * subAge
aP1 <- subaP * subAge
wPAge <- wP1[which(wP1 != 0)]
aPAge <- aP1[which(aP1 != 0)]
t.test(wPAge, aPAge)
```
There is a significant difference.

>Q8. Determine the age of all individuals at time of boost?

```{r}
boostAge <- time_length(ymd(subject$date_of_boost) - ymd(subject$year_of_birth), "years")
boostAge
```

>Q9. With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?

```{r}
subject$age <- boostAge

ggplot(subject) +
  aes(x = age,
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2, ncol =1) 
```

It seems these two groups are significantly different.

There are 3 main datasets in the CMI-PB project at the time of writing,

```{r}
table(subject$dataset)
```

Joining multiple tables

```{r}
# Complete the API URLs...
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/plasma_ab_titer", simplifyVector = TRUE)
```

```{r}
head(specimen)
```

```{r}
head(titer)
```

I want to merge(join) the specimen and subject tables together
 
>Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
#install.packages("dplyr")
library(dplyr)
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

>Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

```{r}
oops <- abdata %>% filter(antigen == "FIM2/3")
table(oops$dataset)
```

```{r}
table(abdata$dataset)
```

>Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

>Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```
Number of visit 8 specimens is very small compared to other visits.

4. Examine IgG1 Ab titer levels

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

>Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:

```{r}
ggplot(ig1) +
  aes(x = MFI, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```

>Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

FIM2/3 showed greatest difference in the level of IgG1 antibody titers recognizing it over time. TT, PRN, FHA and DT also showed differences.

```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```

```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```

>Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
filter(ig1, antigen=="TT") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "TT antigen levels per visit (aP red, wP teal)")
```

```{r}
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "FIM2/3 antigen levels per visit (aP red, wP teal)")
```

>Q16. What do you notice about these two antigens time courses and the FIM2/3 data in particular?

They all appear to peak at around visit 5/6 and then decline. 
FIM2/3 levels clearly rise over time and far exceed those of TT.

>Q17. Do you see any clear difference in aP vs. wP responses?

There is no obvious difference in aP vs. wP responses.

5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

>Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

Select (or filter) for the 2021 dataset and isotype IgG
I want a time course("planned_day_relative_to_boost") of IgG MFI_normalized
for "PT" antigen.

```{r}
#| eval = FALSE
abdata$planned_day_relative_to_boost 
```

```{r}
igpt.21 <- abdata  %>%  
  filter(dataset == "2021_dataset", isotype == "IgG", antigen == "PT")
```

```{r}
ggplot(igpt.21) +
  aes(planned_day_relative_to_boost,
      MFI_normalised,
      col = infancy_vac) +
  geom_point() +
  geom_line(aes(group = subject_id), linewidth = 0.5, alpha = 0.5) +
  geom_smooth(se = FALSE, span = 0.4, linewidth = 3)
```

For 2022 dataset

```{r}
igpt.22 <- abdata  %>%  
  filter(dataset == "2022_dataset", isotype == "IgG", antigen == "PT")
```

```{r}
ggplot(igpt.22) +
  aes(planned_day_relative_to_boost,
      MFI_normalised,
      col = infancy_vac) +
  geom_point() +
  geom_line(aes(group = subject_id), linewidth = 0.5, alpha = 0.5) +
  geom_smooth(se = FALSE, span = 0.4, linewidth = 3) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = 11)
```
```


>Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

It increases and reaches it's maximum level at visit 4 and decreases after that.

>Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

It does not perfectly match the antibody titer trend (maximum at visit 5).
There is a time lag between the increase of IgG1 transcript level and the increase of IgG1 Antibody. Translation takes time.

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

6. Working with larger datasets

```{r}
# Change for your downloaded file path
rnaseq <- read.csv("2020LD_rnaseq.csv")

head(rnaseq,3)
```

```{r}
dim(rnaseq)
```
Working with long format data

```{r}
n_genes <- table(rnaseq$specimen_id)
head( n_genes , 10)
```

How many specimens

```{r}
length(n_genes)
```
Check if there are the same number of genes for each specimen

```{r}
all(n_genes[1]==n_genes)
```
Convert to “wide” format

```{r}
#install.packages("tidyr")
library(tidyr)

rna_wide <- rnaseq %>% 
  select(versioned_ensembl_gene_id, specimen_id, tpm) %>%
  pivot_wider(names_from = specimen_id, values_from=tpm)

dim(rna_wide)
```

```{r}
head(rna_wide[,1:7], 3)
```

